tests/test_qnode.py

```diff
@@ -14,377 +14,553 @@
-        with self.assertRaisesRegex(QuantumFunctionError, 'can only be measured once'):
-            qf(par)
 
-        # device must have enough wires for the qfunc
-        @qml.qnode(self.dev2)
-        def qf(x):
+        node = qml.QNode(circuit, operable_mock_device_2_wires)
+
+        with pytest.raises(QuantumFunctionError, match="can only be measured once"):
+            node(0.5)
+
+    def test_operation_on_nonexistant_wire(self, operable_mock_device_2_wires):
+        """Tests that the QNode properly raises an error if an operation
+           is applied to a non-existant wire."""
+
+        operable_mock_device_2_wires.num_wires = 2
+
+        def circuit(x):
             qml.RX(x, wires=[0])
             qml.CNOT(wires=[0, 2])
             return qml.expval(qml.PauliZ(0))
-        with self.assertRaisesRegex(QuantumFunctionError, 'applied to invalid wire'):
-            qf(par)
 
-        # CV and discrete ops must not be mixed
-        @qml.qnode(self.dev1)
-        def qf(x):
+        node = qml.QNode(circuit, operable_mock_device_2_wires)
+
+        with pytest.raises(QuantumFunctionError, match="applied to invalid wire"):
+            node(0.5)
+
+    def test_observable_on_nonexistant_wire(self, operable_mock_device_2_wires):
+        """Tests that the QNode properly raises an error if an observable
+           is measured on a non-existant wire."""
+
+        operable_mock_device_2_wires.num_wires = 2
+
+        def circuit(x):
+            qml.RX(x, wires=[0])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(2))
+
+        node = qml.QNode(circuit, operable_mock_device_2_wires)
+
+        with pytest.raises(QuantumFunctionError, match="applied to invalid wire"):
+            node(0.5)
+
+    def test_mixing_of_cv_and_qubit_operations(self, operable_mock_device_2_wires):
+        """Tests that the QNode properly raises an error if qubit and
+           CV operations are mixed in the same qfunc."""
+
+        def circuit(x):
             qml.RX(x, wires=[0])
             qml.Displacement(0.5, 0, wires=[0])
             return qml.expval(qml.PauliZ(0))
-        with self.assertRaisesRegex(QuantumFunctionError, 'Continuous and discrete'):
-            qf(par)
 
-        # default plugin cannot execute CV operations, neither gates...
-        @qml.qnode(self.dev1)
-        def qf(x):
-            qml.Displacement(0.5, 0, wires=[0])
-            return qml.expval(qml.X(0))
-        with self.assertRaisesRegex(DeviceError, 'Gate [a-zA-Z]+ not supported on device'):
-            qf(par)
+        node = qml.QNode(circuit, operable_mock_device_2_wires)
 
-        # ...nor observables
-        @qml.qnode(self.dev1)
-        def qf(x):
-            return qml.expval(qml.X(wires=0))
-        with self.assertRaisesRegex(DeviceError, 'Observable [a-zA-Z]+ not supported on device'):
-            qf(par)
+        with pytest.raises(QuantumFunctionError, match="Continuous and discrete"):
+            node(0.5)
```

Comment:
This exception testing class is great, and so much easier to read and debug!

Reply:
Thanks :)